surreal/http.py

```diff
@@ -0,0 +1,124 @@
+"""
+Copyright © SurrealDB Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import base64
+import logging
+from typing import Any
+from urllib.parse import quote
+
+from aiohttp import ClientSession
+
+from surreal.errors import SurrealException
+from surreal.utils import load_json
+
+from . import __version__
+
+_log = logging.getLogger(__name__)
+
+
+def encode_basic_auth(username: str, password: str):
+    username_password = f"{quote(username)}:{quote(password)}"
+
+    return f"Basic {base64.b64encode(username_password.encode()).decode()}"
+
+
+# NOTE: This implements HTTP state and requests
+class HTTPClient:
+    """
+    Raw HTTP Client used to wrap surreal.
+
+    Parameters
+    ----------
+    url: :class:`str`
+        The url your surrealdb instance is running on.
+    username: :class:`str`
+        The username to use to authenticate with your instance.
+    password: :class:`str`
+        The password to use to authenticate with your instance.
+    ns: :class:`str`
+        The namespace on your instance to use.
+    db: :class:`str`
+        The database on your instance to use.
+    """
+    def __init__(
+        self,
+        url: str,
+        username: str,
+        password: str,
+        ns: str,
+        db: str,
+    ) -> None:
+        self.url = f"{url}/sql"
+        self._session: ClientSession | None = None
+
+        encoded_auth = encode_basic_auth(username=username, password=password)
+
+        self._base_headers: dict[str, str] = {
+            "Content-Type": "application/json",
+            "Authorization": f"{encoded_auth}",
+            "User-Agent": f"surreal.py ({__version__})",
+            "NS": ns,
+            "DB": db,
+        }
+
+    async def _create_session(self) -> None:
+        self._session = ClientSession()
```

Comment:
It would make sense to set some defaults here, rather than doing so in every request, for example, the base URL, headers, and authentication can all be set here.